cloud/src/meta-service/meta_service_txn.cpp

```diff
@@ -788,6 +788,164 @@ void update_tablet_stats(const StatsTabletKeyInfo& info, const TabletStats& stat
+            }
+            VersionPB version_pb;
+            if (!version_pb.ParseFromString(ver_val)) {
+                code = MetaServiceCode::PROTOBUF_PARSE_ERR;
+                ss << "failed to parse version pb txn_id=" << txn_id << " key=" << hex(ver_key);
+                msg = ss.str();
+                return;
+            }
+            LOG(INFO) << "txn_id=" << txn_id << " key=" << hex(ver_key)
+                      << " version_pb:" << version_pb.ShortDebugString();
+            DCHECK(version_pb.has_txn_id());
+            DCHECK(version_pb.txn_id() == txn_id);
+            partition_versions.emplace(tmp_rowset_pb.partition_id(), version_pb);
+        }
+
+        const VersionPB& version_pb = partition_versions[tmp_rowset_pb.partition_id()];
+
+        std::string rowset_key =
+                meta_rowset_key({instance_id, tmp_rowset_pb.tablet_id(), version_pb.version()});
+        std::string rowset_val;
+        err = txn->get(rowset_key, &rowset_val);
+        if (TxnErrorCode::TXN_OK == err) {
+            continue;
+        }
+
+        if (err != TxnErrorCode::TXN_KEY_NOT_FOUND) {
+            code = cast_as<ErrCategory::READ>(err);
+            ss << "failed to get rowset_key, txn_id=" << txn_id << " key=" << hex(rowset_key)
+               << " err=" << err;
+            msg = ss.str();
+            LOG(WARNING) << msg;
+            return;
+        }
+        DCHECK(err == TxnErrorCode::TXN_KEY_NOT_FOUND);
+
+        tmp_rowset_pb.set_start_version(version_pb.version());
+        tmp_rowset_pb.set_end_version(version_pb.version());
+
+        rowset_val.clear();
+        if (!tmp_rowset_pb.SerializeToString(&rowset_val)) {
+            code = MetaServiceCode::PROTOBUF_SERIALIZE_ERR;
+            ss << "failed to serialize rowset_meta, txn_id=" << txn_id
+               << " key=" << hex(rowset_key);
+            msg = ss.str();
+            return;
+        }
+
+        txn->put(rowset_key, rowset_val);
+        LOG(INFO) << "xxx put rowset_key=" << hex(rowset_key) << " txn_id=" << txn_id
+                  << " rowset_size=" << rowset_key.size() + rowset_val.size();
+
+        // Accumulate affected rows
+        auto& stats = tablet_stats[tmp_rowset_pb.tablet_id()];
+        stats.data_size += tmp_rowset_pb.data_disk_size();
+        stats.num_rows += tmp_rowset_pb.num_rows();
+        ++stats.num_rowsets;
+        stats.num_segs += tmp_rowset_pb.num_segments();
+    }
+
+    DCHECK(partition_versions.size() == 1);
+
+    for (auto& [tablet_id, stats] : tablet_stats) {
+        DCHECK(tablet_ids.count(tablet_id));
+        auto& tablet_idx = tablet_ids[tablet_id];
+        StatsTabletKeyInfo info {instance_id, tablet_idx.table_id(), tablet_idx.index_id(),
+                                 tablet_idx.partition_id(), tablet_id};
+        update_tablet_stats(info, stats, txn, code, msg);
+        if (code != MetaServiceCode::OK) return;
+    }
+
+    err = txn->commit();
+    if (err != TxnErrorCode::TXN_OK) {
```

Comment:
There may be conflicts caused by concurrent commits. According to our design expectations, TXN_CONFLICT also needs to be handled. One simple solution is to consider the data already converted if TXN_CONFLICT occurs (this requires that each task uses the same split batch method). Another solution is to retry once if TXN_CONFLICT occurs.

It's essential to manage the competition between transactions properly here.